File: instant-quote/src/lib/pricing.ts
// Pure, deterministic pricing engine. No Date, no randomness, no I/O.
// All numbers come from PRICING (pricing-config.ts). Currency: PLN.

import {
  PRICING,
  type LeadTimeId,
  type ProcessId,
  type ProcessDef,
} from './pricing-config'
import type { MeshMetrics } from './mesh/types'
import { countPlates } from './packing'

export type { LeadTimeId, ProcessId } from './pricing-config'

export interface PartConfig {
  process: ProcessId
  quantity: number
  leadTime: LeadTimeId
}

export interface DfmFlag {
  code:
    | 'exceeds_build_volume'
    | 'small_feature'
    | 'min_volume_billed'
    | 'geometry_approximated'
    | 'multi_plate'
  severity: 'block' | 'warn' | 'info'
  message: string
  suggestedProcesses?: ProcessId[]
}

export interface BreakdownLine {
  key: 'material' | 'machine' | 'finishing' | 'plates'
  label: string
  amountPln: number
}

export interface PriceBreak {
  quantity: number
  unitPricePln: number
}

export interface PartQuote {
  blocked: boolean
  billableVolumeCm3: number
  unitBasePln: number
  discountFraction: number
  leadTimeMultiplier: number
  unitPricePln: number
  lineTotalPln: number
  breakdown: BreakdownLine[]
  dfmFlags: DfmFlag[]
  priceBreaks: PriceBreak[]
  // Present only for multi-piece 3MF parts.
  pieceCount?: number
  plates?: number
}

export interface OrderTotals {
  partsSubtotalPln: number
  minOrderTopUpPln: number
  orderFeePln: number
  shippingPln: number
  netTotalPln: number
  vatPln: number
  grossTotalPln: number
  freeShipping: boolean
  minOrderApplied: boolean
}

const round2 = (n: number): number => Math.round(n * 100) / 100

/** Piecewise-linear discount fraction for a quantity, clamped outside tiers. */
export function interpolateDiscount(quantity: number): number {
  const tiers = PRICING.discountTiers
  const q = Math.max(1, Math.floor(quantity))
  if (q <= tiers[0][0]) return tiers[0][1]
  const last = tiers[tiers.length - 1]
  if (q >= last[0]) return last[1]
  for (let i = 0; i < tiers.length - 1; i++) {
    const [lo, loD] = tiers[i]
    const [hi, hiD] = tiers[i + 1]
    if (q >= lo && q <= hi) {
      const t = (q - lo) / (hi - lo)
      return loD + t * (hiD - loD)
    }
  }
  return last[1] // unreachable
}

/**
 * Base unit price for a single part before discount & lead-time multiplier.
 * FDM model (mapi-tech): weight approximates a slicer's walls + solid
 * top/bottom as a shell (surface area × thickness, clamped to the part
 * volume) plus infill of the remaining interior; material cost from that
 * weight × per-material rate and factor. Print time books the shell and
 * infill grams at their own throughputs — perimeters extrude far slower
 * than sparse infill, which is what makes thin-walled parts expensive.
 */
function unitBasePrice(
  proc: ProcessDef,
  volumeCm3: number,
  surfaceAreaCm2: number,
): { total: number; lines: BreakdownLine[] } {
  const shellVolCm3 = Math.min(
    volumeCm3,
    surfaceAreaCm2 * (PRICING.fdm.shellThicknessMm / 10),
  )
  const infillVolCm3 = PRICING.fdm.infillFraction * (volumeCm3 - shellVolCm3)
  const shellG = shellVolCm3 * proc.densityGCm3
  const infillG = infillVolCm3 * proc.densityGCm3
  const weightG = shellG + infillG
  const material = (weightG * proc.plnPerKg * proc.factor) / 1000
  const printH =
    shellG / PRICING.fdm.shellGramsPerPrintHour +
    infillG / PRICING.fdm.infillGramsPerPrintHour
  const machine = printH * proc.plnPerHour
  return {
    total: material + machine,
    lines: [
      { key: 'material', label: 'Material', amountPln: material },
      { key: 'machine', label: 'Machine time', amountPln: machine },
      { key: 'finishing', label: 'Finishing', amountPln: 0 },
    ],
  }
}

/** Sorted descending — lets us compare a part against a build box allowing any rotation. */
function sortedDesc(v: { x: number; y: number; z: number }): number[] {
  return [v.x, v.y, v.z].sort((a, b) => b - a)
}

function fitsBuildVolume(
  proc: ProcessDef,
  bboxMm: MeshMetrics['bboxMm'],
): boolean {
  const part = sortedDesc(bboxMm)
  const build = sortedDesc(proc.build)
  return part[0] <= build[0] && part[1] <= build[1] && part[2] <= build[2]
}

function processesThatFit(bboxMm: MeshMetrics['bboxMm']): ProcessId[] {
  return (Object.keys(PRICING.processes) as ProcessId[]).filter((id) =>
    fitsBuildVolume(PRICING.processes[id], bboxMm),
  )
}

export function computePartQuote(
  metrics: MeshMetrics,
  config: PartConfig,
): PartQuote {
  const proc = PRICING.processes[config.process]
  const dfmFlags: DfmFlag[] = []

  // Billable volume: minimum 1 cm³.
  const rawVolume = metrics.volumeCm3
  const billableVolumeCm3 = Math.max(rawVolume, PRICING.minBillableVolumeCm3)
  if (rawVolume < PRICING.minBillableVolumeCm3) {
    dfmFlags.push({
      code: 'min_volume_billed',
      severity: 'info',
      message: `Under 1 cm³ — billed at the ${PRICING.minBillableVolumeCm3} cm³ minimum.`,
    })
  }

  // Approximated geometry from convex-hull fallback.
  if (metrics.usedHullFallback) {
    dfmFlags.push({
      code: 'geometry_approximated',
      severity: 'warn',
      message:
        'Mesh is not watertight — volume estimated from its convex hull. Final price may change.',
    })
  }

  // Smallest feature warning.
  const minDim = Math.min(metrics.bboxMm.x, metrics.bboxMm.y, metrics.bboxMm.z)
  if (minDim < PRICING.minFeatureMm) {
    dfmFlags.push({
      code: 'small_feature',
      severity: 'warn',
      message: `Smallest dimension is ${minDim.toFixed(2)} mm — thin features may not survive printing.`,
    })
  }

  // Build-volume check for the chosen process (blocking). Multi-piece 3MF
  // files are gated piece-by-piece with plate packing — a merged bbox that
  // exceeds the plate is fine as long as every piece fits (possibly across
  // several plates). Single-piece parts keep the rotation-aware bbox check.
  const pieces = metrics.pieces
  const multiPiece = pieces !== undefined && pieces.length >= 2
  let plates = 1
  let fits: boolean
  if (multiPiece) {
    const packOpts = { gutterMm: PRICING.plateGutterMm }
    const counted = countPlates(pieces, proc.build, packOpts)
    fits = counted !== null
    plates = counted ?? 1
    if (!fits) {
      const alternatives = (
        Object.keys(PRICING.processes) as ProcessId[]
      ).filter(
        (id) =>
          countPlates(pieces, PRICING.processes[id].build, packOpts) !== null,
      )
      dfmFlags.push({
        code: 'exceeds_build_volume',
        severity: 'block',
        message: `A piece exceeds the ${proc.build.x}×${proc.build.y}×${proc.build.z} mm build plate.`,
        suggestedProcesses: alternatives.length ? alternatives : undefined,
      })
    } else if (plates > 1) {
      dfmFlags.push({
        code: 'multi_plate',
        severity: 'info',
        message: `${pieces.length} pieces pack onto ${plates} build plates — ${PRICING.extraPlateFeePln} zł per extra plate.`,
      })
    }
  } else {
    fits = fitsBuildVolume(proc, metrics.bboxMm)
    if (!fits) {
      const alternatives = processesThatFit(metrics.bboxMm)
      dfmFlags.push({
        code: 'exceeds_build_volume',
        severity: 'block',
        message: `Part exceeds the ${proc.build.x}×${proc.build.y}×${proc.build.z} mm build volume.`,
        suggestedProcesses: alternatives.length ? alternatives : undefined,
      })
    }
  }

  const base = unitBasePrice(proc, billableVolumeCm3, metrics.surfaceAreaCm2)
  const baseLines = base.lines
  let unitBasePln = base.total
  // Per-unit fee for each plate beyond the first, folded into the base so
  // discounts, lead-time multipliers, and breakdown scaling apply uniformly.
  const plateFeePln = (plates - 1) * PRICING.extraPlateFeePln
  if (plateFeePln > 0) {
    unitBasePln += plateFeePln
    baseLines.push({
      key: 'plates',
      label: `Extra plates (${plates - 1})`,
      amountPln: plateFeePln,
    })
  }
  const discountFraction = interpolateDiscount(config.quantity)
  const leadTimeMultiplier = PRICING.leadTimes[config.leadTime].mult

  const rawUnit = unitBasePln * (1 - discountFraction) * leadTimeMultiplier
  // mapi-tech floors every part at a minimum price.
  const unitPricePln = Math.max(round2(rawUnit), PRICING.minPartPricePln)
  const qty = Math.max(1, Math.floor(config.quantity))
  const lineTotalPln = round2(unitPricePln * qty)

  // Scale breakdown lines to the line total so they sum exactly.
  const scale = unitBasePln > 0 ? lineTotalPln / unitBasePln : 0
  const scaled = baseLines.map((l) => ({
    ...l,
    amountPln: round2(l.amountPln * scale),
  }))
  // Absorb rounding drift into the machine line so lines sum to lineTotal.
  const drift = lineTotalPln - scaled.reduce((s, l) => s + l.amountPln, 0)
  const machineLine = scaled.find((l) => l.key === 'machine')
  if (machineLine) machineLine.amountPln = round2(machineLine.amountPln + drift)

  // Price-break table for the currently-selected process & lead time.
  const priceBreaks: PriceBreak[] = [1, 5, 10, 25, 50].map((q) => ({
    quantity: q,
    unitPricePln: Math.max(
      round2(unitBasePln * (1 - interpolateDiscount(q)) * leadTimeMultiplier),
      PRICING.minPartPricePln,
    ),
  }))

  return {
    blocked: !fits,
    billableVolumeCm3,
    unitBasePln: round2(unitBasePln),
    discountFraction,
    leadTimeMultiplier,
    unitPricePln,
    lineTotalPln,
    breakdown: scaled,
    dfmFlags,
    priceBreaks,
    ...(multiPiece && { pieceCount: pieces.length, plates }),
  }
}

/**
 * Order-level totals: minimum order, flat order fee, shipping. Blocked parts
 * excluded. All prices are gross (VAT-inclusive, like mapi-tech's checkout);
 * vatPln is the included portion, netTotalPln the gross minus that VAT.
 */
export function computeOrderTotals(quotes: PartQuote[]): OrderTotals {
  const active = quotes.filter((q) => !q.blocked)
  const partsSubtotalPln = round2(
    active.reduce((s, q) => s + q.lineTotalPln, 0),
  )

  const minOrderTopUpPln = round2(
    Math.max(0, PRICING.minOrderPln - partsSubtotalPln),
  )
  const minOrderApplied = minOrderTopUpPln > 0

  const afterMin = round2(partsSubtotalPln + minOrderTopUpPln)
  const orderFeePln = active.length > 0 ? PRICING.orderFeePln : 0
  const freeShipping = afterMin >= PRICING.freeShippingThresholdPln
  const shippingPln = freeShipping ? 0 : PRICING.shippingFlatPln

  const grossTotalPln = round2(afterMin + orderFeePln + shippingPln)
  const vatPln = round2(
    (grossTotalPln * PRICING.vatRate) / (1 + PRICING.vatRate),
  )
  const netTotalPln = round2(grossTotalPln - vatPln)

  return {
    partsSubtotalPln,
    minOrderTopUpPln,
    orderFeePln,
    shippingPln,
    netTotalPln,
    vatPln,
    grossTotalPln,
    freeShipping,
    minOrderApplied,
  }
}

File: backend/internal/pricing/pricing.go
package pricing

import (
	"fmt"
	"math"
	"sort"
	"strconv"
)

type Vec3 struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
	Z float64 `json:"z"`
}

type Piece struct {
	BboxMm Vec3 `json:"bboxMm"`
}

// MeshMetrics is the pricing-relevant subset of the client's mesh analysis.
type MeshMetrics struct {
	VolumeCm3        float64 `json:"volumeCm3"`
	SurfaceAreaCm2   float64 `json:"surfaceAreaCm2"`
	BboxMm           Vec3    `json:"bboxMm"`
	UsedHullFallback bool    `json:"usedHullFallback"`
	Pieces           []Piece `json:"pieces,omitempty"`
}

type PartConfig struct {
	Process  string  `json:"process"`
	Quantity float64 `json:"quantity"`
	LeadTime string  `json:"leadTime"`
}

type DfmFlag struct {
	Code               string   `json:"code"`
	Severity           string   `json:"severity"`
	Message            string   `json:"message"`
	SuggestedProcesses []string `json:"suggestedProcesses,omitempty"`
}

type BreakdownLine struct {
	Key       string  `json:"key"`
	Label     string  `json:"label"`
	AmountPln float64 `json:"amountPln"`
}

type PriceBreak struct {
	Quantity         float64 `json:"quantity"`
	UnitPricePln     float64 `json:"unitPricePln"`
	DiscountFraction float64 `json:"discountFraction"`
}

type PartQuote struct {
	Blocked            bool            `json:"blocked"`
	BillableVolumeCm3  float64         `json:"billableVolumeCm3"`
	UnitBasePln        float64         `json:"unitBasePln"`
	DiscountFraction   float64         `json:"discountFraction"`
	LeadTimeMultiplier float64         `json:"leadTimeMultiplier"`
	UnitPricePln       float64         `json:"unitPricePln"`
	LineTotalPln       float64         `json:"lineTotalPln"`
	Breakdown          []BreakdownLine `json:"breakdown"`
	DfmFlags           []DfmFlag       `json:"dfmFlags"`
	PriceBreaks        []PriceBreak    `json:"priceBreaks"`
	// Present only for multi-piece 3MF parts.
	PieceCount *int `json:"pieceCount,omitempty"`
	Plates     *int `json:"plates,omitempty"`
}

type OrderTotals struct {
	PartsSubtotalPln float64 `json:"partsSubtotalPln"`
	MinOrderTopUpPln float64 `json:"minOrderTopUpPln"`
	OrderFeePln      float64 `json:"orderFeePln"`
	ShippingPln      float64 `json:"shippingPln"`
	NetTotalPln      float64 `json:"netTotalPln"`
	VatPln           float64 `json:"vatPln"`
	GrossTotalPln    float64 `json:"grossTotalPln"`
	FreeShipping     bool    `json:"freeShipping"`
	MinOrderApplied  bool    `json:"minOrderApplied"`
}

// round2 replicates JS `Math.round(n*100)/100`: half-up toward +Inf,
// which differs from Go's math.Round (half away from zero) for negatives.
func round2(n float64) float64 {
	return math.Floor(n*100+0.5) / 100
}

// fmtNum renders a float the way a JS template literal does: shortest
// round-trip decimal ("340", "0.2", "1.5").
func fmtNum(v float64) string {
	return strconv.FormatFloat(v, 'f', -1, 64)
}

// InterpolateDiscount is the piecewise-linear discount fraction for a
// quantity, clamped outside the tiers.
func (c *Config) InterpolateDiscount(quantity float64) float64 {
	tiers := c.DiscountTiers
	q := math.Max(1, math.Floor(quantity))
	if q <= tiers[0].Quantity {
		return tiers[0].Fraction
	}
	last := tiers[len(tiers)-1]
	if q >= last.Quantity {
		return last.Fraction
	}
	for i := 0; i < len(tiers)-1; i++ {
		lo, loD := tiers[i].Quantity, tiers[i].Fraction
		hi, hiD := tiers[i+1].Quantity, tiers[i+1].Fraction
		if q >= lo && q <= hi {
			t := (q - lo) / (hi - lo)
			return loD + t*(hiD-loD)
		}
	}
	return last.Fraction // unreachable
}

// unitBasePrice: shell (surface area × thickness, clamped to volume) plus
// infill of the remaining interior; material from weight × per-kg rate and
// factor; machine time books shell and infill grams at separate throughputs.
func (c *Config) unitBasePrice(proc ProcessDef, volumeCm3, surfaceAreaCm2 float64) (float64, []BreakdownLine) {
	shellVolCm3 := math.Min(volumeCm3, surfaceAreaCm2*(c.Fdm.ShellThicknessMm/10))
	infillVolCm3 := c.Fdm.InfillFraction * (volumeCm3 - shellVolCm3)
	shellG := shellVolCm3 * proc.DensityGCm3
	infillG := infillVolCm3 * proc.DensityGCm3
	weightG := shellG + infillG
	material := (weightG * proc.PlnPerKg * proc.Factor) / 1000
	printH := shellG/c.Fdm.ShellGramsPerPrintHour + infillG/c.Fdm.InfillGramsPerPrintHour
	machine := printH * proc.PlnPerHour
	return material + machine, []BreakdownLine{
		{Key: "material", Label: "Material", AmountPln: material},
		{Key: "machine", Label: "Machine time", AmountPln: machine},
		{Key: "finishing", Label: "Finishing", AmountPln: 0},
	}
}

// sortedDesc lets us compare a part against a build box allowing any rotation.
func sortedDesc(v Vec3) [3]float64 {
	d := [3]float64{v.X, v.Y, v.Z}
	sort.Sort(sort.Reverse(sort.Float64Slice(d[:])))
	return d
}

func fitsBuildVolume(proc ProcessDef, bboxMm Vec3) bool {
	part := sortedDesc(bboxMm)
	build := sortedDesc(Vec3(proc.Build))
	return part[0] <= build[0] && part[1] <= build[1] && part[2] <= build[2]
}

func (c *Config) processesThatFit(bboxMm Vec3) []string {
	var ids []string
	for _, p := range c.Processes {
		if fitsBuildVolume(p, bboxMm) {
			ids = append(ids, p.ID)
		}
	}
	return ids
}

func (c *Config) ComputePartQuote(metrics MeshMetrics, config PartConfig) PartQuote {
	proc, _ := c.Process(config.Process)
	dfmFlags := []DfmFlag{}

	// Billable volume: minimum 1 cm³.
	rawVolume := metrics.VolumeCm3
	billableVolumeCm3 := math.Max(rawVolume, c.MinBillableVolumeCm3)
	if rawVolume < c.MinBillableVolumeCm3 {
		dfmFlags = append(dfmFlags, DfmFlag{
			Code:     "min_volume_billed",
			Severity: "info",
			Message:  fmt.Sprintf("Under 1 cm³ — billed at the %s cm³ minimum.", fmtNum(c.MinBillableVolumeCm3)),
		})
	}

	// Approximated geometry from convex-hull fallback.
	if metrics.UsedHullFallback {
		dfmFlags = append(dfmFlags, DfmFlag{
			Code:     "geometry_approximated",
			Severity: "warn",
			Message:  "Mesh is not watertight — volume estimated from its convex hull. Final price may change.",
		})
	}

	// Smallest feature warning.
	minDim := math.Min(metrics.BboxMm.X, math.Min(metrics.BboxMm.Y, metrics.BboxMm.Z))
	if minDim < c.MinFeatureMm {
		dfmFlags = append(dfmFlags, DfmFlag{
			Code:     "small_feature",
			Severity: "warn",
			Message:  fmt.Sprintf("Smallest dimension is %s mm — thin features may not survive printing.", strconv.FormatFloat(minDim, 'f', 2, 64)),
		})
	}

	// Build-volume check for the chosen process (blocking). Multi-piece 3MF
	// files are gated piece-by-piece with plate packing; single-piece parts
	// keep the rotation-aware bbox check.
	pieces := metrics.Pieces
	multiPiece := len(pieces) >= 2
	plates := 1
	var fits bool
	if multiPiece {
		counted := CountPlates(pieces, proc.Build, PackOpts{GutterMm: c.PlateGutterMm})
		fits = counted != nil
		if counted != nil {
			plates = *counted
		}
		if !fits {
			var alternatives []string
			for _, p := range c.Processes {
				if CountPlates(pieces, p.Build, PackOpts{GutterMm: c.PlateGutterMm}) != nil {
					alternatives = append(alternatives, p.ID)
				}
			}
			dfmFlags = append(dfmFlags, DfmFlag{
				Code:     "exceeds_build_volume",
				Severity: "block",
				Message: fmt.Sprintf("A piece exceeds the %s×%s×%s mm build plate.",
					fmtNum(proc.Build.X), fmtNum(proc.Build.Y), fmtNum(proc.Build.Z)),
				SuggestedProcesses: alternatives,
			})
		} else if plates > 1 {
			dfmFlags = append(dfmFlags, DfmFlag{
				Code:     "multi_plate",
				Severity: "info",
				Message: fmt.Sprintf("%d pieces pack onto %d build plates — %s zł per extra plate.",
					len(pieces), plates, fmtNum(c.ExtraPlateFeePln)),
			})
		}
	} else {
		fits = fitsBuildVolume(proc, metrics.BboxMm)
		if !fits {
			alternatives := c.processesThatFit(metrics.BboxMm)
			dfmFlags = append(dfmFlags, DfmFlag{
				Code:     "exceeds_build_volume",
				Severity: "block",
				Message: fmt.Sprintf("Part exceeds the %s×%s×%s mm build volume.",
					fmtNum(proc.Build.X), fmtNum(proc.Build.Y), fmtNum(proc.Build.Z)),
				SuggestedProcesses: alternatives,
			})
		}
	}

	unitBasePln, baseLines := c.unitBasePrice(proc, billableVolumeCm3, metrics.SurfaceAreaCm2)
	// Per-unit fee for each plate beyond the first, folded into the base so
	// discounts, lead-time multipliers, and breakdown scaling apply uniformly.
	plateFeePln := float64(plates-1) * c.ExtraPlateFeePln
	if plateFeePln > 0 {
		unitBasePln += plateFeePln
		baseLines = append(baseLines, BreakdownLine{
			Key:       "plates",
			Label:     fmt.Sprintf("Extra plates (%d)", plates-1),
			AmountPln: plateFeePln,
		})
	}
	discountFraction := c.InterpolateDiscount(config.Quantity)
	leadTime, _ := c.LeadTime(config.LeadTime)
	leadTimeMultiplier := leadTime.Mult

	rawUnit := unitBasePln * (1 - discountFraction) * leadTimeMultiplier
	// mapi-tech floors every part at a minimum price.
	unitPricePln := math.Max(round2(rawUnit), c.MinPartPricePln)
	qty := math.Max(1, math.Floor(config.Quantity))
	lineTotalPln := round2(unitPricePln * qty)

	// Scale breakdown lines to the line total so they sum exactly.
	scale := 0.0
	if unitBasePln > 0 {
		scale = lineTotalPln / unitBasePln
	}
	scaled := make([]BreakdownLine, len(baseLines))
	for i, l := range baseLines {
		l.AmountPln = round2(l.AmountPln * scale)
		scaled[i] = l
	}
	// Absorb rounding drift into the machine line so lines sum to lineTotal.
	sum := 0.0
	for _, l := range scaled {
		sum += l.AmountPln
	}
	drift := lineTotalPln - sum
	for i := range scaled {
		if scaled[i].Key == "machine" {
			scaled[i].AmountPln = round2(scaled[i].AmountPln + drift)
			break
		}
	}

	// Price-break table for the currently-selected process & lead time.
	priceBreaks := make([]PriceBreak, 0, len(QuantityChips))
	for _, q := range QuantityChips {
		d := c.InterpolateDiscount(float64(q))
		priceBreaks = append(priceBreaks, PriceBreak{
			Quantity:         float64(q),
			UnitPricePln:     math.Max(round2(unitBasePln*(1-d)*leadTimeMultiplier), c.MinPartPricePln),
			DiscountFraction: d,
		})
	}

	quote := PartQuote{
		Blocked:            !fits,
		BillableVolumeCm3:  billableVolumeCm3,
		UnitBasePln:        round2(unitBasePln),
		DiscountFraction:   discountFraction,
		LeadTimeMultiplier: leadTimeMultiplier,
		UnitPricePln:       unitPricePln,
		LineTotalPln:       lineTotalPln,
		Breakdown:          scaled,
		DfmFlags:           dfmFlags,
		PriceBreaks:        priceBreaks,
	}
	if multiPiece {
		n := len(pieces)
		p := plates
		quote.PieceCount = &n
		quote.Plates = &p
	}
	return quote
}

// ComputeOrderTotals: minimum order, flat order fee, shipping. Blocked parts
// excluded. All prices are gross (VAT-inclusive); vatPln is the included
// portion, netTotalPln the gross minus that VAT.
func (c *Config) ComputeOrderTotals(quotes []PartQuote) OrderTotals {
	var active []PartQuote
	for _, q := range quotes {
		if !q.Blocked {
			active = append(active, q)
		}
	}
	sum := 0.0
	for _, q := range active {
		sum += q.LineTotalPln
	}
	partsSubtotalPln := round2(sum)

	minOrderTopUpPln := round2(math.Max(0, c.MinOrderPln-partsSubtotalPln))
	minOrderApplied := minOrderTopUpPln > 0

	afterMin := round2(partsSubtotalPln + minOrderTopUpPln)
	orderFeePln := 0.0
	if len(active) > 0 {
		orderFeePln = c.OrderFeePln
	}
	freeShipping := afterMin >= c.FreeShippingThresholdPln
	shippingPln := c.ShippingFlatPln
	if freeShipping {
		shippingPln = 0
	}

	grossTotalPln := round2(afterMin + orderFeePln + shippingPln)
	vatPln := round2((grossTotalPln * c.VatRate) / (1 + c.VatRate))
	netTotalPln := round2(grossTotalPln - vatPln)

	return OrderTotals{
		PartsSubtotalPln: partsSubtotalPln,
		MinOrderTopUpPln: minOrderTopUpPln,
		OrderFeePln:      orderFeePln,
		ShippingPln:      shippingPln,
		NetTotalPln:      netTotalPln,
		VatPln:           vatPln,
		GrossTotalPln:    grossTotalPln,
		FreeShipping:     freeShipping,
		MinOrderApplied:  minOrderApplied,
	}
}
